app/experimenter/experiments/models.py

```diff
@@ -804,6 +807,60 @@ def json_dumps_value(self):
         return json.dumps(json.loads(self.value), indent=2)
 
 
+class VariantPreferences(models.Model):
+    variant = models.ForeignKey(
+        ExperimentVariant,
+        blank=False,
+        null=False,
+        related_name="preferences",
+        on_delete=models.CASCADE,
+    )
+    pref_name = models.CharField(max_length=255, blank=False, null=False)
+    pref_type = models.CharField(
+        max_length=255,
+        choices=ExperimentConstants.PREF_TYPE_CHOICES,
+        blank=True,
+        null=True,
+    )
+    pref_branch = models.CharField(
+        max_length=255,
+        choices=ExperimentConstants.PREF_BRANCH_CHOICES,
+        blank=True,
+        null=True,
+    )
+
+    pref_value = models.CharField(max_length=255, blank=False, null=False)
+
+    class Meta:
+        unique_together = (("variant", "pref_name"),)
+
+    def clean(self):
+        super().clean()
+        expected_type_mapping = expected_type = {
+            Experiment.PREF_TYPE_BOOL: bool,
+            Experiment.PREF_TYPE_INT: int,
+            Experiment.PREF_TYPE_STR: str,
+        }
+        expected_type = expected_type_mapping.get(self.pref_type, None)
+        if self.pref_type != Experiment.PREF_TYPE_STR:
+            try:
+                found_type = type(json.loads(self.pref_value))
+
+                # type validation only for non json type
+                if expected_type and found_type != expected_type:
+                    raise ValueError
+
+            except (json.JSONDecodeError, ValueError):
+                raise ValidationError(
+                    "value", f"Unexpected value type (should be {self.pref_type})"
+                )
+
+    def save(self, *args, **kwargs):
+        self.clean()
+
+        super().save()
```

Comment:
This is the same as a regular variant, can't we pull this out into a mixin?

Reply:
Actually do we need to do this on the model at all?

Reply:
I think the rationale is that VariantPreferences are saved separately from variants, and this was put in place just in case validation from the serializer fails?

Reply:
Note: we chatted about this on zoom and agreed to do validation in the serializers and not on the model.